lib/core/scheduler.rb

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'time'
+require 'fugit'
+require 'core/connector_settings'
+require 'utility/cron'
+require 'utility/logger'
+require 'utility/exception_tracking'
+
+module Core
+  class Scheduler
+    def initialize(connector_id, poll_interval)
+      @connector_id = connector_id
+      @poll_interval = poll_interval
+    end
+
+    def when_triggered
+      loop do
+        connector_settings = Core::ConnectorSettings.fetch(@connector_id)
+
+        if sync_triggered?(connector_settings)
+          yield connector_settings
+        end
+      rescue StandardError => e
+        Utility::ExceptionTracking.log_exception(e, 'Sync failed due to unexpected error.')
+      ensure
+        if @poll_interval > 0
+          Utility::Logger.info("Sleeping for #{@poll_interval} seconds.")
+          sleep(@poll_interval)
+        end
+      end
+    end
+
+    private
+
+    def sync_triggered?(connector_settings)
+      unless connector_settings.connector_status_allows_sync?
+        Utility::Logger.info("Connector #{connector_settings.id} is in status \"#{connector_settings.connector_status}\" and won't sync yet. Connector needs to be in one of the following statuses: #{Connectors::ConnectorStatus::STATUSES_ALLOWING_SYNC} to run.")
```

Comment:
we starting to have a lot of messages here. I would put them in constants at the top of the file to keep the code easy to read

Reply:
Good call, I'll make a follow-up PR, though it'll change the usage of logger a bit to style of `printf "Connector %s is in status '%s'", connector_settings.id, connector_settings.status` - which is pretty okay too.